### PR TITLE
Upgrade CodeQualityExtension to use lazy properties

### PR DESCRIPTION
**Pull request by asodja**

### CODE REVIEW

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle clean

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test AST

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

**Review by tresat — APPROVED** [review]

File: platforms/jvm/code-quality/src/main/java/org/gradle/api/plugins/quality/CheckstyleExtension.java [hunk 1]
```diff
@@ -46,6 +46,7 @@ public abstract class CheckstyleExtension extends CodeQualityExtension {
     private final Property<Boolean> enableExternalDtdLoad;
 
     public CheckstyleExtension(Project project) {
+        super();
```
[the remark this answers is not in the document]
Comment (asodja) [reply platforms/jvm/code-quality/src/main/java/org/gradle/api/plugins/quality/CheckstyleExtension.java:49]: Thanks! Removed

File: platforms/jvm/code-quality/src/main/java/org/gradle/api/plugins/quality/CodeNarcExtension.java [hunk 2]
```diff
@@ -42,6 +42,7 @@ public abstract class CodeNarcExtension extends CodeQualityExtension {
     private String reportFormat;
 
     public CodeNarcExtension(Project project) {
+        super();
```
[the remark this answers is not in the document]
Comment (asodja) [reply platforms/jvm/code-quality/src/main/java/org/gradle/api/plugins/quality/CodeNarcExtension.java:45]: Removed

File: platforms/jvm/code-quality/src/main/java/org/gradle/api/plugins/quality/internal/AbstractCodeQualityPlugin.java [hunk 3]
```diff
@@ -213,17 +212,9 @@ private void configureCheckTaskDependents() {
...
-                        return Iterables.transform(extension.getSourceSets(), new Function<SourceSet, String>() {
-                            @Override
-                            public String apply(SourceSet sourceSet) {
-                                return sourceSet.getTaskName(taskBaseName, null);
-                            }
-                        });
-                    }
-                });
+                task.dependsOn(extension.getSourceSets().map(sourceSets ->
+                    Iterables.transform(sourceSets, sourceSet -> sourceSet.getTaskName(taskBaseName, null))
```
[the remark this answers is not in the document]
Comment (asodja) [reply platforms/jvm/code-quality/src/main/java/org/gradle/api/plugins/quality/internal/AbstractCodeQualityPlugin.java:216]: Fixed

File: platforms/jvm/code-quality/src/main/java/org/gradle/api/plugins/quality/PmdExtension.java [hunk 4]
```diff
@@ -44,6 +44,7 @@ public abstract class PmdExtension extends CodeQualityExtension {
     private final Property<Integer> threads;
 
     public PmdExtension(Project project) {
+        super();
```
[the remark this answers is not in the document]
Comment (asodja) [reply platforms/jvm/code-quality/src/main/java/org/gradle/api/plugins/quality/PmdExtension.java:47]: Removed

File: build-logic/documentation/src/main/groovy/gradlebuild/docs/GradleJavadocsPlugin.java [hunk 5]
```diff
@@ -146,10 +148,20 @@ private void generateJavadocs(Project project, ProjectLayout layout, TaskContain
...
             task.setConfig(project.getResources().getText().fromFile(checkstyle.getConfigDirectory().file("checkstyle-api.xml")));
             task.setClasspath(layout.files());
-            task.getReports().getXml().getOutputLocation().set(new File(checkstyle.getReportsDir(), "checkstyle-api.xml"));
+            task.getReports().getXml().getOutputLocation().set(getCheckstyleOutputLocation(checkstyle, objects));
         });
     }
 
+    @SuppressWarnings({"ConstantValue", "CastCanBeRemovedNarrowingVariableType"})
+    private static Provider<RegularFile> getCheckstyleOutputLocation(CheckstyleExtension checkstyle, ObjectFactory objects) {
+        Object reportsDir = checkstyle.getReportsDir();
```
[the remark this answers is not in the document]
Comment (asodja) [reply build-logic/documentation/src/main/groovy/gradlebuild/docs/GradleJavadocsPlugin.java:157]: It depends with which Gradle version it's run: 
if it's using wrapper version, then it's a file. But after upgrade this becomes `DirectoryProperty` and Gradleception tests fail. So I had to add this workaround that handles both cases

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

**Review by tresat — APPROVED** [review]

**Pull request merged**